components/TextFields/src/ContainedInputView/MDCBaseTextFieldLayout.m

```diff
@@ -0,0 +1,120 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "MDCBaseTextFieldLayout.h"
+
+static const CGFloat kHorizontalPadding = (CGFloat)12.0;
+
+/**
+ This constant can go away once the code that determines container height is brought over.
+ */
+static const CGFloat kTemporaryVerticalPadding = (CGFloat)12.0;
+
+@interface MDCBaseTextFieldLayout ()
+@end
+
+@implementation MDCBaseTextFieldLayout
+
+#pragma mark Object Lifecycle
+
+- (instancetype)initWithTextFieldSize:(CGSize)textFieldSize
+                             leftView:(UIView *)leftView
+                         leftViewMode:(UITextFieldViewMode)leftViewMode
+                            rightView:(UIView *)rightView
+                        rightViewMode:(UITextFieldViewMode)rightViewMode
+                                isRTL:(BOOL)isRTL
+                            isEditing:(BOOL)isEditing {
```

Comment:
Can we provide a reference to the text field here instead of each of these parameters? This is a private API, so that level of direct access may prove easier to work with than the strict separation of concerns here.

Reply:
I don't like everything about the way it is here--the initializer does become comically long once you account for everything. At the same time, I think the separation of concerns keeps the object from morphing into something it shouldn't be. It keeps us honest! And I like that this file doesn't have to import the textfield itself.

Also, a lot of the stuff that gets passed into this initializer is private on the TextField, so just passing in the TextField would require the maintenance of `+Private` categories, which I don't really want to deal with. Can we leave this one the way it is? 😅

Reply:
Did you consider using a protocol to capture the required shape of the text field? Using a protocol would meet most of the goals you described here while also simplifying the usage of your layout object.

Reply:
E.g.:

```
// private/MDCBaseTextFieldLayout.h
@protocol MDCBaseTextFieldLayoutContext
- (CGSize)textFieldSize;
...
@end

Reply:
I think I did consider creating a layout "context" or "config" object or something like that, and moving the really long initializer to that, but ultimately didn't see a benefit in doing so.

Reply:
It doesn't look like `init` is unavailable, so I would guess this verbose initializer is only needed for clients who want something custom.   I like the idea of `- init` for *typical use* and `- initWithAll:The:Things:` for other uses.  Just my two cents.

Reply:
It may help readability quite a bit while also allowing you to document the behavior of each property more directly.

Reply:
@romoore, This is intended to be a private class, so I'm not too worried about clients misusing it. I could add an `NS_UNAVAILABLE` to `-init` and an `NS_DESIGNATED_INITIALIZER` to the big long one though.

@jverkoey, I could document the initializer with a whole bunch of things like `@param leftView the TextField's leftView property`, `@param isRTL A BOOL that specifies whether the TextField is in an RTL locale`, etc.?

Reply:
One good thing about this big long initializer is there's really no way to do it wrong. If you just called `-init` and set a bunch of properties you might forget one. The compiler won't let you forget one in this case.

All the properties should be publicly readonly also. The fact that they're not is an oversight.

Reply:
Looking at https://github.com/material-components/material-components-ios/blob/develop/components/TextFields/examples/experimental/supplemental/InputChipViewLayout.m, storing and referencing a single context object defined by a protocol will significantly increase readability of this layout object's implementation.

For example, this block of code

https://github.com/material-components/material-components-ios/blob/a1d1296884a3158ed1c32c04cd437ee2b11ef278/components/TextFields/examples/experimental/supplemental/InputChipViewLayout.m#L160-L196

becomes instead:

```objc
  NSArray<NSValue *> *chipFrames = [self chipFrames];

  CGRect textFieldFrame = [self textFieldFrame];
  CGPoint contentOffset = [self scrollViewContentOffset];
  CGSize contentSize = [self scrollViewContentSize];
```

Reply:
I see how the 4 line bottom snippet is more alluring than the longer input chip view (not textfield) layout snippet above it, but I'm not convinced it's any more scalable. I'm a little concerned it's the opposite. In the 4 line bottom snippet is `textFieldFrame` supposed to be another property on the layout object that's set before the `contentOffset` and `contentSize` methods referencing it are called? This would have to be the case in order for it to work, but it wouldn't necessarily be obvious. It would be very unclear which elements of the layout depended on which other elements. It would be possible to call things out of order, where it currently is not. Right now these layout objects are executed in a top-down fashion, and the helper methods (most of which haven't been brought over yet) can be counted on to return the right values because those values are only based on the parameters passed in rather than the `self`'s configurable-by-anyone state.

I understand that it's a small headache dealing with these parameter lists but I still think that for this particular case it's the lesser of many evils.

Reply:
Thanks for replying, @andrewoverton I didn't see the file path for this class. Given that it's a private implementation detail, it seems like it could easily be refactored later, especially if the same grouping of parameters are reused across APIs.  Also given the amount of back-and-forth on this particular point, it may be worth bringing this discussion to the team.  What do you think, @jverkoey ?

Reply:
Agreed @romoore; I'll be grabbing Andrew to chat today 1:1 to clarify and then resolve this thread.

Reply:
Spoke offline and agreed to move forward with this code as-is while keeping an eye out for potential improvements.